### PR TITLE
Make preload scripts take a function declaration instead of expression

### PR DESCRIPTION
**Pull request by jgraham**
This is a prerequisite to adding support for arguments, which is required to allow passing in message channels.

This also cleans up a few bugs with evaluating functions.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/pull/381.html" title="Last updated on Mar 30, 2023, 11:05 AM UTC (bc976d4)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver-bidi/381/99db2d4...bc976d4.html" title="Last updated on Mar 30, 2023, 11:05 AM UTC (bc976d4)">Diff</a>

### CODE REVIEW

**Comment by sadym-chromium** [comment]
Handing over to @thiagowfx

**Comment by whimboo** [comment]
FYI related test changes could be done via https://bugzilla.mozilla.org/show_bug.cgi?id=1822779

File: index.bs [hunk 1]
```diff
@@ -4779,8 +4799,8 @@ The [=remote end steps=] given |session| and |command parameters| are:
 
 1. Let |preload script map| be |session|'s [=preload script map=].
 
-1. Set |preload script map|[|script|] to a struct with <code>expression</code>
-   |expression| and <code>sandbox</code> |sandbox|.
+1. Set |preload script map|[|script|] to a struct with <code>function
+   declaration</code> |function declaration|, , and <code>sandbox</code> |sandbox|.
```
Comment (whimboo) [inline index.bs:4803]: I think we want to use the `<code>` variant here, and please also remove the left over comma.

**Review by whimboo — CHANGES_REQUESTED** [review]
Thanks @jgraham for those changes! I had a first pass over the code and it looks fine to me in general. I have a couple of questions and comments through. Please have a look at those.

**Comment by whimboo** [comment]
I think that we should get this out sooner than later before clients actually start using this API and would be confronted with backward incompatible changes.

File: index.bs (hunk 1 above)
Comment (jgraham) [inline index.bs:4803]:
```suggestion
   declaration</code> |function declaration|, and <code>sandbox</code> |sandbox|.
```

Comment (jgraham) [reply index.bs]: I don't know what you mean by "the `<code>` variant".

File: index.bs [hunk 2]
```diff
@@ -4769,7 +4789,7 @@ script=].
 <div algorithm="remote end steps for script.addPreloadScript">
 The [=remote end steps=] given |session| and |command parameters| are:
 
-1. Let |expression| be the <code>expression</code> field of |command
+1. Let |function declaration| be the <code>function declaration</code> field of |command
```
Comment (juliandescottes) [inline index.bs:4792]:
```suggestion
1. Let |function declaration| be the <code>functionDeclaration</code> field of |command
```

nit: Should be `functionDeclaration` here?

File: index.bs (hunk 1 above)
Comment (whimboo) [reply index.bs]: I've read that wrongly. Please ignore.

**Comment by thiagowfx** [comment]
> Handing over to @thiagowfx

Seems OK to me. Keep in mind I am just starting to get familiar with the spec though.

**Review by thiagowfx — APPROVED** [review]

File: index.bs [hunk 3]
```diff
@@ -4225,24 +4231,36 @@ To <dfn export>run WebDriver BiDi preload scripts</dfn> given |environment setti
...
+       1. [=Report the exception=] |error|.
+
+    1. [=Prepare to run script=] with |environment settings|.
+
+    1. Set |evaluation status| to
+       [=Call=](|function object|, null, «»).
 
     1. [=Clean up after running script=] with |environment settings|.
 
+    1. If |function body evaluation status| is an [=abrupt completion=], then
```
Comment (whimboo) [inline index.bs:4261]:
```suggestion
    1. If |evaluation status| is an [=abrupt completion=], then
```

**Review by whimboo — CHANGES_REQUESTED** [review]

File: index.bs [hunk 4]
```diff
@@ -4225,24 +4231,36 @@ To <dfn export>run WebDriver BiDi preload scripts</dfn> given |environment setti
...
+       1. Let |error| be a new [=TypeError=] object in |realm|.
 
-    1. If |evaluation status| is an [=abrupt completion=], then [=report the
-       exception=] given by |evaluation status|.\[[Value]] for |script|.
+       1. [=Report the exception=] |error|.
+
+    1. [=Prepare to run script=] with |environment settings|.
+
+    1. Set |evaluation status| to
+       [=Call=](|function object|, null, «»).
```
[the remark this answers is not in the document]
Comment (jgraham) [reply index.bs:4258]: It's an empty list, as used in Infra and ECMAScript. See https://infra.spec.whatwg.org/#lists. Undefined would be wrong.

File: index.bs [hunk 5]
```diff
@@ -4225,24 +4230,38 @@ To <dfn export>run WebDriver BiDi preload scripts</dfn> given |environment setti
...
+    1. Let |function object| be |function body evaluation status|.\[[Value]].
+
+    1. If [=IsCallable=](|function object|) is <code>false</code>:
 
-    1. Let |evaluation status| be [=ScriptEvaluation=](|script|'s record).
+       1. Let |error| be a new [=TypeError=] object in |realm|.
 
-    1. If |evaluation status| is an [=abrupt completion=], then [=report the
-       exception=] given by |evaluation status|.\[[Value]] for |script|.
+       1. [=Report the exception=] |error|.
```
Comment (whimboo) [inline index.bs:4253]: Oh, I missed to ask if this error should actually be a dedicated WebDriver specific error or if `TypeError` is fine here which later would be converted to a `UnknownError` most likely.

**Review by whimboo — APPROVED** [review]
Looks fine to me now. Thanks @jgraham! Just a nit inline you might want to have a look.

File: index.bs (hunk 5 above)
Comment (jgraham) [reply index.bs:4253]: It doesn't ever get converted into a WebDriver error; there's no command in progress at the point at which preload scripts are run. Instead an exception is reported to the console and via `log.logEntryAdded` if the client is subscribed. We _could_ theoretically define some new JS error types for exceptions that happen in WebDriver scripts, but I don't think that's a good idea (UAs can already do things like replace the script filename by something that indicates a preload script in the stack trace).

**Pull request merged**